### PR TITLE
[ncp] add border routing InfraIf send ICMP6 ND

### PR DESCRIPTION
**Pull request by Irving-cl**
This PR implements sending InfraIf Icmp6 Nd messages on NCP.

This is to support border routing for NCP case. See this https://github.com/openthread/ot-br-posix/issues/2398#issuecomment-2348518919 for the whole picture of this series of changes.

In this PR, a new spinel property SPINEL_PROP_INFRA_IF_SEND_ICMP6 is added for the NCP to send the ICMP6 messages to the host.

### CODE REVIEW

File: src/lib/spinel/spinel.h [hunk 1]
```diff
@@ -4763,6 +4763,16 @@ enum
      */
     SPINEL_PROP_INFRA_IF_RECV_ICMP6 = SPINEL_PROP_INFRA_IF__BEGIN + 2,
 
+    /// ICMP6 message sent by NCP and needs to be sent on the infrastructure interface.
+    /** Format: `L6d`
+     * Type: Unsolicited notifications only
+     *
+     * `L`: The infrastructure interface index.
+     * `6`: The IP6 source address of the message to send.
```
Comment (zhanglongxia) [inline src/lib/spinel/spinel.h:4771]: The function `InfraIfSendIcmp6Nd()` sets the IPv6 address as the destination address. Should this address be the destination address?

Comment (Irving-cl) [reply src/lib/spinel/spinel.h]: Good catch! Updated.

File: src/ncp/platform/infra_if.cpp [hunk 2]
```diff
@@ -41,12 +41,11 @@ otError otPlatInfraIfSendIcmp6Nd(uint32_t            aInfraIfIndex,
...
 {
-    OT_UNUSED_VARIABLE(aInfraIfIndex);
-    OT_UNUSED_VARIABLE(aDestAddress);
-    OT_UNUSED_VARIABLE(aBuffer);
-    OT_UNUSED_VARIABLE(aBufferLength);
+    otError           error = OT_ERROR_NONE;
+    ot::Ncp::NcpBase *ncp   = ot::Ncp::NcpBase::GetNcpInstance();
 
-    return OT_ERROR_NOT_IMPLEMENTED;
+    error = ncp->InfraIfSendIcmp6Nd(aInfraIfIndex, aDestAddress, aBuffer, aBufferLength);
```
Comment (abtink) [inline src/ncp/platform/infra_if.cpp:47]: How about `return ncp->InfraIfSendIcmp6Nd(...)`?

**Review by abtink — APPROVED** [review]
👍

File: src/ncp/platform/infra_if.cpp (hunk 2 above)
Comment (Irving-cl) [reply src/ncp/platform/infra_if.cpp]: Updated!

**Review by zhanglongxia — APPROVED** [review]
LGTM

File: src/ncp/ncp_base.hpp [hunk 3]
```diff
@@ -224,6 +224,27 @@ class NcpBase
...
+     *                            platform should do the checksum calculate.
+     * @param[in]  aBufferLength  The length of the message buffer.
+     *
+     * @note  Per RFC 4861, the implementation should send the message with IPv6 link-local source address
+     *        of interface @p aInfraIfIndex and IP Hop Limit 255.
+     *
+     * @retval OT_ERROR_NONE    Successfully sent the ICMPv6 message.
+     * @retval OT_ERROR_FAILED  Failed to send the ICMPv6 message.
+     *
+     */
```
Comment (jwhui) [inline src/ncp/ncp_base.hpp:242]: nit:
```suggestion
     * @retval OT_ERROR_FAILED  Failed to send the ICMPv6 message.
     */
```

**Review by jwhui — APPROVED** [review]

**Pull request merged**